test/reduce.jl

```diff
@@ -31,6 +31,8 @@ using .Main.OffsetArrays
 @test Base.mapfoldr(abs2, -, 2:5) == -14
 @test Base.mapfoldr(abs2, -, 2:5; init=10) == -4
 
+@test foldr((x, y) -> ('⟨' * x * '|' * y * '⟩'), "λ 🐨.α") == "⟨λ|⟨ |⟨🐨|⟨.|α⟩⟩⟩⟩" # issue #31780
+
```

Comment:
Perhaps it is worth adding `@allocated` test for the `foo` and `bar` functions (in https://github.com/JuliaLang/julia/pull/31781#issuecomment-485164510) since getting those to not allocate was a bit finicky?